### PR TITLE
DDS-715 auth service affiliates

### PR DESCRIPTION
**Pull request by dmlond**
implements:
- get /auth_services/:id/affiliates
- get /auth_services/:id/affiliates/:uid
- post /auth_services/:id/affiliates/:uid/dds_user

adds concept of affiliate_search to auth_providers
auth_providers support affiliate_search if they belong_to an identity_provider

affiliate search endpoints return an exception if the auth_service does not support affiliate_search

### CODE REVIEW

**Comment by dmlond** [comment]
I need to add rake tasks for identity_provider management. Hold off on the review

File: app/api/dds/v1/auth_providers_api.rb
Comment (dmann) [inline app/api/dds/v1/auth_providers_api.rb]: I think we're going to have to figure out a way to paginate these results at some point.

Comment (dmann) [reply app/api/dds/v1/auth_providers_api.rb]: Since this will be a breaking change, should we paginate before merging?

File: app/models/identity_provider.rb
Comment (dmann) [inline app/models/identity_provider.rb:1]: Since `IdentityProvider` is an interface, I think it would be good to add `affiliate` and `affiliates` methods that raise NotImplementedError, to ensure that child classes implement them.

File: app/models/ldap_identity_provider.rb
Comment (dmann) [inline app/models/ldap_identity_provider.rb]: I think `ldap_search` should be private.

File: spec/factories/users.rb
Comment (dmann) [inline spec/factories/users.rb]: `Faker::Name.first_name` includes names with apostrophes, like D'Angelo, which can lead to transient test failures.  :(

File: spec/models/ldap_identity_provider_spec.rb
Comment (dmann) [inline spec/models/ldap_identity_provider_spec.rb]: You can get rid of the following context by replacing the above with:
```
it { is_expected.to respond_to(:affiliate).with(1).argument }
```

Comment (dmann) [reply spec/models/ldap_identity_provider_spec.rb]: https://www.relishapp.com/rspec/rspec-expectations/docs/built-in-matchers/respond-to-matcher

File: spec/models/ldap_identity_provider_spec.rb
Comment (dmann) [inline spec/models/ldap_identity_provider_spec.rb]: Nit picky, but please prepend method names with `#` for instance and `::` for class methods.

File: spec/models/ldap_identity_provider_spec.rb
Comment (dmann) [inline spec/models/ldap_identity_provider_spec.rb]: Please prepend method name with `#`

File: spec/support/ldap_shared_examples.rb
Comment (dmann) [inline spec/support/ldap_shared_examples.rb:15]: Cool! I'm sure there are other places in our code that could use this technique.

**Review by dmann — CHANGES_REQUESTED** [review]
Looks good for the most part. Just asking for a few minor changes.

File: app/api/dds/v1/auth_providers_api.rb
Comment (dmlond) [reply app/api/dds/v1/auth_providers_api.rb]: go ahead and merge. We will need to work through a cycle of customer feedback, Matt specifying API design for pagination, developer spiking LDAP pagination and implementation as a new feature.

**Comment by dmlond** [comment]
all but pagination are addressed

File: app/api/dds/v1/auth_providers_api.rb
Comment (dmann) [reply app/api/dds/v1/auth_providers_api.rb]: @Donmurry what do you think?

Comment (Donmurry) [reply app/api/dds/v1/auth_providers_api.rb]: Hmm, I think it might be best to add pagination in now, if possible.  If spike of LDAP pagination is going to take weeks vs days, then i say we can skip it.  But if we can incorporate it now, I think thats better than not.  We can weigh the cost of development now vs cost later.

Comment (dmann) [reply app/api/dds/v1/auth_providers_api.rb]: If the decision is to skip pagination, it would be good to inform the client developers of this decision and the reasoning behind it. I think they would appreciate a warning that, if they submit a query with 3 characters, they're likely to get a firehose of data OR at request timeout. That way when both of those outcomes eventually happen, hopefully, they won't make an "issue" out of it...

Comment (dmlond) [reply app/api/dds/v1/auth_providers_api.rb]: https://github.com/ruby-ldap/ruby-net-ldap/blob/e4c46a223a19feda78393a793711353aa1febdcd/lib/net/ldap/connection.rb

I really dont think the ruby net-ldap search interface allows you access to pagination. They paginate the results under the hood, but they repeatedly search until all results are returned before returning to the caller with all the results.  They do allow you to send a maximum size, and time value (this might be useful to prevent timeouts), but these just end up minimizing each call to get the next page's results, while still continuously getting each pages results until they are all returned.

If we want to support pagination, we will have to do it wastefully, e.g. get all results, skip to N, add until M, break. This would satisfy clients requirement if they end up needing pagination, but would not prevent server timeouts of memory explosions.

Comment (dmann) [reply app/api/dds/v1/auth_providers_api.rb]: I feel strongly that this endpoint needs to limit the number of returned results. The pagination default is 100, with a max of 1000. Ideally, the client would have the ability change this limit via the `per_page` parameter.

While it would be nice to do a full implementation of pagination, I don't believe that is necessary. That is, I don't think that we need to allow clients to use the `page` parameter. In fact, if `page` is provided, I suggest returning a "Not Implemented" or "Not supported by Identity Provider" error of some kind.

This is a technical limitation of ldap and that shouldn't dictate how our interface works. Let's say that you initially setup the affiliates endpoints with an identity provider that supported pagination and you were easily able to add support for `per_page` and `page`. Feature released, done. Now, you have to add a second identity provider that does not support pagination. What do you do? Remove pagination from the endpoint? Probably not. This is the way that I'm approaching this problem and hopefully provides context to my suggestions above.

**Review by dmann — APPROVED** [review]

**Pull request merged**
